zookeeper-api/src/test/java/org/apache/helix/zookeeper/impl/client/TestDedicatedZkClient.java

```diff
@@ -0,0 +1,154 @@
+import org.apache.helix.zookeeper.datamodel.ZNRecord;
+import org.apache.helix.zookeeper.datamodel.serializer.ZNRecordSerializer;
+import org.apache.helix.zookeeper.impl.factory.DedicatedZkClientFactory;
+import org.apache.helix.zookeeper.impl.factory.MetadataStoreRoutingData;
+import org.apache.helix.zookeeper.impl.factory.TrieRoutingData;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+
+public class TestDedicatedZkClient extends ZkTestBase {
+  private static final String ZK_SHARDING_KEY_PREFIX = "/TEST_SHARDING_KEY_";
+  private static final String TEST_VALID_PATH = ZK_SHARDING_KEY_PREFIX + "_" + 0 + "/a/b/c";
+  private static final String TEST_INVALID_PATH = ZK_SHARDING_KEY_PREFIX + "_invalid" + "/a/b/c";
+
+  // <Realm, List of sharding keys> Mapping
+  private static final Map<String, List<String>> RAW_ROUTING_DATA = new HashMap<>();
+
+  private RealmAwareZkClient _realmAwareZkClient;
+  private MetadataStoreRoutingData _metadataStoreRoutingData;
+
+  @BeforeClass
+  public void beforeClass()
+      throws Exception {
+    // Populate RAW_ROUTING_DATA
+    for (int i = 0; i < _numZk; i++) {
+      List<String> shardingKeyList = new ArrayList<>();
+      shardingKeyList.add(ZK_SHARDING_KEY_PREFIX + "_" + i);
+      String realmName = ZK_PREFIX + (ZK_START_PORT + i);
+      RAW_ROUTING_DATA.put(realmName, shardingKeyList);
+    }
+
+    // Feed the raw routing data into TrieRoutingData to construct an in-memory representation of routing information
+    _metadataStoreRoutingData = new TrieRoutingData(RAW_ROUTING_DATA);
+  }
+
+  /**
+   * 1. Create a DedicatedZkClient with a non-existing sharding key (for which there is no valid ZK realm)
+   * -> This should fail with an exception
+   * 2. Create a DedicatedZkClient with a valid sharding key
+   * -> This should pass
+   */
+  @Test
+  public void testDedicatedZkClientCreation() {
+    // Create a DedicatedZkClient
+    String invalidShardingKey = "InvalidShardingKey";
+    RealmAwareZkClient.RealmAwareZkClientConfig clientConfig =
+        new RealmAwareZkClient.RealmAwareZkClientConfig();
+
+    // Create a connection config with the invalid sharding key
+    RealmAwareZkClient.RealmAwareZkConnectionConfig connectionConfig =
+        new RealmAwareZkClient.RealmAwareZkConnectionConfig(invalidShardingKey);
+
+    try {
+      _realmAwareZkClient = DedicatedZkClientFactory.getInstance()
+          .buildZkClient(connectionConfig, clientConfig, _metadataStoreRoutingData);
+      Assert.fail("Should not succeed with an invalid sharding key!");
+    } catch (IllegalArgumentException e) {
+      // Expected
+    }
+
+    // Use a valid sharding key this time around
+    String validShardingKey = ZK_SHARDING_KEY_PREFIX + "_" + 0; // Use TEST_SHARDING_KEY_0
+    connectionConfig = new RealmAwareZkClient.RealmAwareZkConnectionConfig(validShardingKey);
+    _realmAwareZkClient = DedicatedZkClientFactory.getInstance()
```

Comment:
Is this client supposed to be close after class?

Reply:
Yes. Will close.